src/daemon/ca/publishing.rs

```diff
@@ -1,232 +1,1223 @@
-            old,
+pub struct CaObjects {
+    ca: Handle,
+    repo: Option<RepositoryContact>,
+
+    #[serde(with = "ca_objects_classes_serde")]
+    classes: HashMap<ResourceClassName, ResourceClassObjects>,
+
+    #[serde(skip_serializing_if = "Vec::is_empty", default)]
+    deprecated_repos: Vec<RepositoryContact>,
+}
+
+mod ca_objects_classes_serde {
+
+    use super::*;
+
+    use serde::de::{Deserialize, Deserializer};
+    use serde::ser::Serializer;
+    #[derive(Debug, Deserialize)]
+    struct ClassesItem {
+        class_name: ResourceClassName,
+        objects: ResourceClassObjects,
+    }
+
+    #[derive(Debug, Serialize)]
+    struct ClassesItemRef<'a> {
+        class_name: &'a ResourceClassName,
+        objects: &'a ResourceClassObjects,
+    }
+
+    pub fn serialize<S>(
+        map: &HashMap<ResourceClassName, ResourceClassObjects>,
+        serializer: S,
+    ) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.collect_seq(
+            map.iter()
+                .map(|(class_name, objects)| ClassesItemRef { class_name, objects }),
+        )
+    }
+
+    pub fn deserialize<'de, D>(deserializer: D) -> Result<HashMap<ResourceClassName, ResourceClassObjects>, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let mut map = HashMap::new();
+        for item in Vec::<ClassesItem>::deserialize(deserializer)? {
+            map.insert(item.class_name, item.objects);
         }
+        Ok(map)
     }
+}
 
-    pub fn name(&self) -> &ObjectName {
-        &self.name
+impl CaObjects {
+    pub fn new(
+        ca: Handle,
+        repo: Option<RepositoryContact>,
+        classes: HashMap<ResourceClassName, ResourceClassObjects>,
+    ) -> Self {
+        CaObjects {
+            ca,
+            repo,
+            classes,
+            deprecated_repos: vec![],
+        }
     }
 
-    pub fn current(&self) -> &CurrentObject {
-        &self.current
+    #[allow(clippy::clippy::mutable_key_type)]
+    pub fn elements(&self) -> HashMap<RepositoryContact, Vec<PublishElement>> {
+        let mut res = HashMap::new();
+
+        if let Some(repo) = &self.repo {
+            res.insert(repo.clone(), vec![]);
+
+            for rco in self.classes.values() {
+                rco.add_elements(&mut res, repo);
+            }
+        }
+
+        res
     }
 
-    pub fn next_update(&self) -> Time {
-        self.next_update
+    pub fn ca_take_deprecated_repos(&mut self) -> Vec<RepositoryContact> {
+        let mut res = vec![];
+        std::mem::swap(&mut res, &mut self.deprecated_repos);
+        res
+    }
+
+    /// Add a new resource class, this returns an error in case the class already exists.
+    fn add_class(
+        &mut self,
+        class_name: &ResourceClassName,
+        key: &CertifiedKey,
+        timing: &IssuanceTimingConfig,
+        signer: &KrillSigner,
+    ) -> KrillResult<()> {
+        if self.classes.contains_key(class_name) {
```

Comment:
Not sure if this would work or help with performance or if you consider this less readabile, but an alternative might be `self.classes.entry(class_name).or_insert_with(...)` (though I'm not sure how that works with the `class_name.clone()`...).